contracts/p1/RToken.sol

```diff
@@ -178,79 +182,159 @@ contract RTokenP1 is ComponentP1, ERC20PermitUpgradeable, IRToken {
+        emit Redemption(_msgSender(), recipient, amount, basketsRedeemed);
+        emit BasketsNeededChanged(basketsNeeded, basketsNeeded - basketsRedeemed);
+        basketsNeeded = basketsNeeded - basketsRedeemed;
+
+        (address[] memory erc20s, uint256[] memory amounts) = basketHandler.quote(
+            basketsRedeemed,
+            FLOOR
+        );
+
+        // === Interactions ===
+
+        for (uint256 i = 0; i < erc20s.length; ++i) {
+            if (amounts[i] == 0) continue;
+
+            // Send withdrawal
+            IERC20Upgradeable(erc20s[i]).safeTransferFrom(
+                address(backingManager),
+                recipient,
+                amounts[i]
+            );
+        }
+    }
+
+    /// Redeem RToken for a linear combination of historical baskets, to a particular recipient
+    // checks:
+    //   amount > 0
+    //   amount <= balanceOf(caller)
+    //   sum(portions) == FIX_ONE
+    //   nonce >= basketHandler.primeNonce() for nonce in basketNonces
+    //
+    // effects:
+    //   (so totalSupply -= amount and balanceOf(caller) -= amount)
+    //   basketsNeeded' / totalSupply' >== basketsNeeded / totalSupply
     //   burn(caller, amount)
+    //
+    // actions:
     //   for each token in erc20s:
-    //     let tokenAmt = (amount * basketsNeeded / totalSupply) baskets of support for token
+    //     let tokenAmt = (amount * basketsNeeded / totalSupply) custom baskets
     //     let prorataAmt = (amount / totalSupply) * token.balanceOf(backingManager)
     //     do token.transferFrom(backingManager, caller, min(tokenAmt, prorataAmt))
     // untestable:
     //      `else` branch of `exchangeRateIsValidAfter` (ie. revert)
     //       BU exchange rate cannot decrease, and it can only increase when < FIX_ONE.
+    /// @dev Allows partial redemptions up to the minAmounts
     /// @param recipient The address to receive the backing collateral tokens
     /// @param amount {qRTok} The quantity {qRToken} of RToken to redeem
-    /// @param basketNonce The nonce of the basket the redemption should be from; else reverts
-    /// @custom:interaction
-    function redeemTo(
+    /// @param basketNonces An array of basket nonces to do redemption from
+    /// @param portions {1} An array of Fix quantities that must add up to FIX_ONE
+    /// @param erc20sOut An array of ERC20s expected out
+    /// @param minAmounts {qTok} The minimum ERC20 quantities the caller should receive
+    /// @custom:interaction RCEI
+    function redeemToCustom(
         address recipient,
         uint256 amount,
-        uint48 basketNonce
-    ) public notFrozen exchangeRateIsValidAfter {
+        uint48[] memory basketNonces,
+        uint192[] memory portions,
+        IERC20[] memory erc20sOut,
+        uint256[] memory minAmounts
+    ) external notFrozen exchangeRateIsValidAfter {
         // == Refresh ==
+
         assetRegistry.refresh();
+        main.furnace().melt();
```

Comment:
So, this one is interesting. Previously we had a try/catch because the modifier on the Furnace was `notPausedOrFrozen`, while redeem() was `notFrozen`. Now the modifier on the Furnace is `notFrozen`, because melting is required (i) in issuance, and (ii) during trading. So it _should_ be the case that we don't need a try/catch. But we're also talking redemption here! It would be really bad, for example, to be able to brick redemption by swapping out the Furnace maliciously. So I think I'll throw the try/catch back in there, for safety. Good catch.